Fix one of the tests for a remote storage QueueManager

dmilstein opened the pull request:
Specifically, the TestSpawnNotMoreThanMaxConcurrentSendsGoroutines was failing on a fresh checkout of master (go 1.7 on OS X, fwiw).

The test had a race condition -- it would only pass if one of the spawned goroutines happened to very quickly pull a set of samples off an internal queue.

This patch rewrites the test so that it deterministically waits until all samples have been pulled off that queue.  In case of errors, it also now reports on the difference between what it expected and what it found.

I verified that, if the code under test is broken, the test successfully reports on that.

@beorn7

grobie commented on storage/remote/queue_manager_test.go:
```diff
@@ -129,20 +141,33 @@ func TestSpawnNotMoreThanMaxConcurrentSendsGoroutines(t *testing.T) {
+	// extra goroutines, but something totally unexpected), we deadlock, which
+	// is a clear signal of a problem.
+	for len(m.queue) > 0 {
+		time.Sleep(10 * time.Millisecond)
```
Hum, I'd read more about the code first. But should we have a timeout for that, so that if things go completely wrong, we abort after something like 1s or whatever makes sense?

beorn7 replied:
Tests are always running with a timeout anyway, but it's 10m by default. So bailing out here after 100-something iterations sounds like a good plan.

dmilstein commented:
Nice points, added an exit after 100 iterations = 1 second (plus an informative message).

Also fixed up something with the deferred closes so that the test doesn't get hung up in error cases.

grobie commented:
:+1:

beorn7 commented:
:+1: Thanks a bunch.
